### ISSUE TITLE
Support user-configured env

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#11061]
**What would you like Renovate to be able to do?**

Allow users to specify literal env variables to be used.

**Did you already have any implementation ideas?**

I suggest we can use a new config option `env` for this, and it should be a simple string:string object.

Example from #7361 for Go: `GOPROXY`, `GONOPROXY`, `GOPRIVATE`, `GONOSUMDB`.

What makes this complex is that these values could be used by both the `go` datasource as well as the `gomod` manager artifacts updating. Ideally we always allow granular configuration, e.g. certain `env` values for `gomod` but different ones for `npm` or `dockerfile`. In that case we can't "write" them to `process.env` before extraction and instead need to *pass* them to both the manager functions as well as datasource lookups. A further challenge is that we have a goal that datasources *don't* return different results per-repo, otherwise they cannot be cached.

So what if `env` instead is restricted be a "top level" repo config option, and we *do* "write" the values to `process.env` prior to extraction, assuming they satisfy some type of allow list we have to protect ourselves? It would still mean we could get some cache problems if the values differ between repos.

**Comment by rarkins** [comment]
Other examples of where env could be used: Dockerfiles, or `.npmrc` substitution

**Comment by secustor** [comment]
IMHO there should definitly be a whitelist of allowed variable keys as this can be for example be used to proxy communication of a hosted app runner through a abitrary proxy.   

Further I'm not sure if it is intended that app users could enable experimental features this way.

**Comment by rarkins** [comment]
We could have one or more "safe" prefixes we define. I doubt `RENOVATE_` is a good idea considering we use that prefix ourselves. But could e.g. be `RENOVATE_ENV_`, `CUSTOM_ENV_`, `USER_ENV_`. Then naturally we should let bot admins define patterns they allow apart from the default we suggest.

**Comment by gmariottiAtWolt** [comment]
Joining the issue from https://github.com/renovatebot/renovate/discussions/20473. Currently, one of the issues we have with Gradle lockfile and the Github App, is that it's not possible to pass credentials for a private Gradle/Maven repository to the process that updates the lockfile. For self-hosted renovate, `customEnvVariables` works perfectly.

**Comment by rarkins** [comment]
But would it be solved if renovate passes your env to child processes once this new feature was implemented?

**Comment by gmariottiAtWolt** [comment]
I think so. For simplicity, however, if you decide to restrict the variables to the ones with a specific prefix, it would be great to have the option to keep or remove that prefix. Using Gradle as an example, defining a custom repository can be done with

```kotlin
repositories {
    maven {
        name = "myArtifactory"
        credentials(PasswordCredentials::class)
        url = uri("https://my-artifactory.org")
    }
}
```

And the necessary credentials passed with the variables `ORG_GRADLE_PROJECT_myArtifactoryUsername` and `ORG_GRADLE_PROJECT_myArtifactoryPassword`. This avoids the need to explicitly define `username` and `password` in the `maven` block using `System.env` and custom env variables names

**Comment by rarkins** [comment]
Is ORG_GRADLE_PROJECT a static string or something that is dynamically configurable?

**Comment by nikolaik** [comment]
Would love a way to use in repo renovate.json configuration to allow the Renovate github app to opt in to experimental behaviour like the features listed here https://docs.renovatebot.com/self-hosted-experimental I'm particularly interested in `RENOVATE_X_GITHUB_HOST_RULES`. 

My thinking could be a common interface to check if experiments are enabled or not, that can be enabled either using env vars or using a config option like `experiments` which is a list of enabled feature toggles/experiments names, ie.
```json
"experiments": ["github_host_rules"]
```
would be equivalent to
```sh
export RENOVATE_X_GITHUB_HOST_RULES=1
```

**Comment by rarkins** [comment]
I would like to increase the priority of this one.

Today we have `customEnvVariables` which per the docs:
> This configuration will be applied after all other environment variables so you can use it to override defaults.

So it's not only for additional env but can be used in theory to fix wrong env which Renovate might already be passing.

If users are allowed to specify env in repo config, should that then be applied _after_ the existing custom env variables? I think that makes most sense, especially as the bot admin can control the allowed ones.

Next challenge is how to make the allow list practical. It's kind of painful to need to specify every single env, when 99% of them are safe. Perhaps we have a user-curated list in a preset can be added to globalExtends so that the majority of users can use that.

Configuration-wise, we can add `allowedEnv` to global config and simply `env` to repo config options? Then we need to decide whether to keep `customEnvVariables` as-is or rename it to match. If matching then we need to work out a way to ensure global custom env does not need allowedEnv in order to work.

**Comment by rarkins** [comment]
Conclusion:
- Add `allowedEnv` to global config. It should be an array of strings and use our glob or regex matching
- Add `env` as a repo-configurable config (key/value strings)
- Config validation should check env keys against allowedEnv and raise config warning issue if there's a problem
- `env` should be merged after `customEnvVariables` when calling child processes. We should iterate over it and compare to `allowedEnv` _just in case_ and raise a warning if we need to filter out any (it implies that we missed validating them in config validation)
- Check for non-child process use cases like GOPROXY and we need to decide if/how to handle it
- A `globalExtends` preset can be added later (not in initial PR)

**Comment by RahulGautamSingh** [comment]
>Note to developer: check in code for non-child process use cases of process.env like GOPROXY and we need to decide if/how to handle it

What does non-child mean in this context?

**Comment by rarkins** [comment]
There should be no changes to child process needed

**Comment by rarkins** [comment]
Model it on existing custom env option

**Comment by RahulGautamSingh** [comment]
That is a global option and we access it using the `GlobalConfig.get` method, whereas this neeeds to passed down in functions as arguments so there will a difference.

**Comment by RahulGautamSingh** [comment]
> There should be no changes to child process needed

We need to pass the user configured env variables till here right
https://github.com/renovatebot/renovate/blob/cffb3d256a9583fbcb3f646903d8dd9904917919/lib/util/exec/utils.ts#L20-L25

**Comment by rarkins** [comment]
Yes, the variables should be merged prior to it reaching child process.

They would need to be passed to exec(). The ExecOptions type has env, extraEnv. When I comment out `env` it doesn't seem to have wide use (only within lib/util).

So perhaps we should use extraEnv. In this case we need to make sure `env` is passed through to commands like updateArtifacts(), probably through UpdateArtifactsConfig

**Comment by rarkins** [comment]
Actually let's keep it simple: add new option `userEnv` to `ExecOptions` and pass the `config.env` into that. Then it's a bit more transparent which env comes from users and which from our managers own logic.

**Comment by RahulGautamSingh** [comment]
Can we do this in 2 PRs? One which focuses on the child processes and the other which focuses on datasources and any remaining non child processes. I feel the amount of changes will be hard/time-consuming to review and discuss in a single PR.

**Comment by RahulGautamSingh** [comment]
Regarding the non-child processes, I think the implementation for `go` datasoruce is pretty simple ie. pass the value to getReleases function and consume them in the following functions:

https://github.com/renovatebot/renovate/blob/8706b771233ca8008687e6a93878fe1999076189/lib/modules/datasource/go/releases-goproxy.ts#L119

https://github.com/renovatebot/renovate/blob/8706b771233ca8008687e6a93878fe1999076189/lib/modules/datasource/go/releases-goproxy.ts#L188-L190

https://github.com/renovatebot/renovate/blob/8706b771233ca8008687e6a93878fe1999076189/lib/modules/datasource/go/releases-goproxy.ts#L345-L347


Apart from `go` the other use-cases which are discussedd above, ie. `dockerfile` and `npmrc`. 
 - In the datasource `docker` we do use the `RENOVATE_X_DOCKER_MAX_PAGES` and `RENOVATE_X_DOCKER_HUB_TAGS`,
    Do we want to configure these vars using the `userEnv` or you meant something else?
    
 - In the npmrc logic,  I think we will need to add new logic here to substitute the variables from npmrc with userEnv values here:
 https://github.com/renovatebot/renovate/blob/6269932d0b99c76ef3a8b86f7070ca064b99d3bb/lib/modules/datasource/npm/npmrc.ts#L150-L154

**Comment by viceice** [comment]
`renovate_x` (and all other experimental environment variables) are admin only

**Comment by rarkins** [comment]
> Regarding the non-child processes, I think the implementation for `go` datasoruce is pretty simple ie. pass the value to getReleases function and consume them in the following functions:

Does this mean that we need to add `env` to our `GetReleasesConfig`?

> Apart from `go` the other use-cases which are discussedd above, ie. `dockerfile` and `npmrc`.
> 
> * In the datasource `docker` we do use the `RENOVATE_X_DOCKER_MAX_PAGES` and `RENOVATE_X_DOCKER_HUB_TAGS`,
>   Do we want to configure these vars using the `userEnv` or you meant something else?

We never want to allow RENOVATE_X_ user env to be taken into consideration, even if the admin allows it as a pattern.

> * In the npmrc logic,  I think we will need to add new logic here to substitute the variables from npmrc with userEnv values here:
>   https://github.com/renovatebot/renovate/blob/6269932d0b99c76ef3a8b86f7070ca064b99d3bb/lib/modules/datasource/npm/npmrc.ts#L150-L154

This is probably a valid case where the user `env` should be taken into consideration, although I'm not sure if anyone will need it.

**Comment by RahulGautamSingh** [comment]
> Does this mean that we need to add `env` to our `GetReleasesConfig`?

 Yes

**Comment by RahulGautamSingh** [comment]
@rarkins Should we close this one and open new issue when UserEnv is required for non-child process tasks?
